### PR TITLE
Fixed #35717 -- Reduced Create/RemoveCollation operations when optimizing migrations.

### PR DESCRIPTION
**Pull request by adamchainz**
#### Trac ticket number

ticket-35717

#### Branch description

Add `CreateCollation.reduce()` which will optimize away the addition + removal of a collation.

#### Checklist
- [x] This PR targets the `main` branch. 
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [x] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

File: tests/postgres_tests/test_operations.py [hunk 1]
```diff
@@ -459,6 +459,24 @@ def test_writer(self):
...
+                    "sample_collation",
+                    "und-u-ks-level2",
+                    provider="icu",
+                    deterministic=False,
+                ),
+                RemoveCollation(
+                    "sample_collation",
+                    # Different locale
+                    "de-u-ks-level1",
+                ),
```
Comment (charettes) [inline tests/postgres_tests/test_operations.py:475]: I assume that holds true because Postgres doesn't allow you to have collations with different locales sharing the same name? Kind of two difference indices with difference parameters that are included in this case to allow the operation to be reversible, [something we missed with `RemoveField`](https://code.djangoproject.com/ticket/26739).

File: django/db/migrations/operations/fields.py [hunk 2]
```diff
@@ -134,7 +134,7 @@ def reduce(self, operation, app_label):
         ):
             if isinstance(operation, AlterField):
                 return [
-                    AddField(
+                    self.__class__(
```
Comment (charettes) [inline django/db/migrations/operations/fields.py:137]: Are these done in order to support a particular third-party application use case? I don't see any tests to make sure we don't regress here though.

**Review by charettes — COMMENTED** [review]
The changes look good but I'm not sure I entirely understand the motive behind the intermediate commit's `self.__class__` change. I could see why it might be useful but it'd be great to have some minimal testing around it.

**Comment by adamchainz** [comment]
> The changes look good but I'm not sure I entirely understand the motive behind the intermediate commit's `self.__class__` change. I could see why it might be useful but it'd be great to have some minimal testing around it.

That commit is from #18510, to fix the reduction bug.

**Comment by sarahboyce** [comment]
The first commit has been merged in f5ddd54986172c29c32f5d835584ea237d5a3781
This can be rebased :+1:

File: tests/postgres_tests/test_operations.py (hunk 1 above)
Comment (adamchainz) [reply tests/postgres_tests/test_operations.py:475]: > I assume that holds true because Postgres doesn't allow you to have collations with different locales sharing the same name?

Yes, exactly.

> Kind of two difference indices with difference parameters that are included in this case to allow the operation to be reversible, something we missed with RemoveField.

Yes, `RemoveCollation` can already be reversible because it accepts all parameters.

**Comment by adamchainz** [comment]
Rebased. This can now be reviewed and merged separately to #18510.

**Review by felixxm — APPROVED** [review]
@adamchainz Thanks :+1:

File: tests/postgres_tests/test_operations.py [hunk 3]
```diff
@@ -459,6 +459,24 @@ def test_writer(self):
...
+            ],
+            [],
+        )
+
```
Comment (sarahboyce) [inline tests/postgres_tests/test_operations.py:479]: Should the reverse (they remove and add it back in) also reduce to nothing?
This fails currently
```suggestion
        self.assertOptimizesTo(
            [
                RemoveCollation("sample_collation", locale="C"),
                CreateCollation("sample_collation", locale="C"),
            ],
            [],
        )
```

Comment (felixxm) [reply tests/postgres_tests/test_operations.py:479]: As far as I'm aware, this cannot happen in a valid operations list. You cannot remove something that was not created in advance.

Comment (sarahboyce) [reply tests/postgres_tests/test_operations.py:479]: Confirmed this isn't something we support with other things like `RemoveField` then `AddField`

The scenario I was thinking was squashing some migrations in an app so I made a model with a field, makemigrations (`0001` creates the model with a field), remove field makemigrations (`0002` removes a field), add it back makemigrations (`0002` adds the field back) 
run `python manage.py squashmigrations myapp 0002 0003` and it says it can't optimize (I expect it can reduce to nothing). Optimizes as expected when doing `python manage.py squashmigrations myapp 0003`

So, long story short, this is separate to this PR and there might be reasons why we've decided not to support this

Comment (felixxm) [reply tests/postgres_tests/test_operations.py:479]: IMO, we cannot optimize such scenario because there is no guarantee that the field created in `0003` is the same as the field created in `0001`.

**Review by sarahboyce — APPROVED** [review]
Thank you all 👍

File: tests/postgres_tests/test_operations.py (hunk 3 above)
Comment (felixxm) [reply tests/postgres_tests/test_operations.py:479]: For example:
- `0001` - `CreateCollation("sample_collation", locale="C"),`
- `0002` - `RemoveCollation("sample_collation", locale="C")`
- `0003` - `CreateCollation("sample_collation", locale="und-u-ks-level2", provider="icu", deterministic=False)`

We can optimize `0001`-`0002` to no-op, but we cannot optimize `0002`-`0003` to no-op as we would end with `"sample_collation", locale="C"`.

Comment (adamchainz) [reply tests/postgres_tests/test_operations.py:479]: Exactly. And even if they are the same, migrations 0001-0003 can be squashed together, at which point the optimizer would remove the create/remove pair.
